### PR TITLE
Move mesh generation to vispy

### PR DESCRIPTION
**Pull request by kcpevey**
# Description

I've added a benchmark to check for regressions when switching to updating the mesh with every slice event. 

Based on my understanding, calling `refresh` should trigger both `on_data_change` _and_ `set_view_slice`. This means that `refresh` should encompass the mesh code in both locations and be as close to apples to apples as possible. 

That said, I'm not certain whether `refresh` should be called more than once in order to capture the additional burden of recalculating the mesh with every slice. I believe that in the old code, there was basically a check to see if anything had changed and if not, the mesh wasn't recalculated. This would mean that running refresh multiple times on the old code should be less computationally expensive than the new code _unless_ the reduction to only generate the mesh in the slice subset is sufficiently efficient. 

TLDR - I benchmarked against calling refresh in the test function once, and again with three consecutive calls to refresh.

**RESULTS:**

**call refresh once**
```
**old code**
[100.00%] ··· ======== =============
               param1               
              -------- -------------
                 16     2.81±0.02ms 
                 64     4.67±0.09ms 
                256      10.2±0.1ms 
                1024     36.3±0.8ms 
                4096     74.6±0.2ms 
               16384      110±1ms   
               65536      250±3ms   
              ======== =============


**new code**
[100.00%] ··· ======== =============
               param1               
              -------- -------------
                 16     2.93±0.03ms 
                 64     4.70±0.06ms 
                256      10.5±0.1ms 
                1024     38.1±0.9ms 
                4096     75.8±0.4ms 
               16384     122±0.8ms  
               65536      298±2ms   
              ======== =============
```

**call refresh 3 times**
```
**old code**
[100.00%] ··· ======== =============
               param1               
              -------- -------------
                 16      8.30±0.1ms 
                 64     13.9±0.07ms 
                256      31.2±0.3ms 
                1024     105±0.6ms  
                4096      216±1ms   
               16384      326±6ms   
               65536      756±9ms   
              ======== =============

**new code**
[100.00%] ··· ======== =============
               param1               
              -------- -------------
                 16     8.63±0.02ms 
                 64     14.3±0.09ms 
                256      32.0±0.2ms 
                1024     106±0.7ms  
                4096      224±2ms   
               16384     358±0.8ms  
               65536      896±4ms   
              ======== =============
```   

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

I added this benchmark as a qt benchmark because otherwise, the benchmark wasn't getting into `on_data_change`. It was strange because when I ran the same commands through pytest, I could have gotten away with not instantiating the viewer (I think). If there is a better approach, I'm happy to hear it! 

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/benchmarks/benchmark_qt_viewer_vectors.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        np.random.seed(0)
+        self.data = np.random.random((n, 2, 3))
+        self.viewer = napari.Viewer()
+        self.layer = self.viewer.add_vectors(self.data)
+        self.visual = self.viewer.window._qt_viewer.layer_to_visual[self.layer]
+
+    def teardown(self, n):
+        self.viewer.window.close()
+
+    def time_vectors_refresh(self, n):
```
Comment (sofroniewn) [inline napari/benchmarks/benchmark_qt_viewer_vectors.py:27]: Could you split this out into two functions - one `time_vectors_refresh` with just a single call and one `time_vectors_multi_refresh`.

**Review by sofroniewn — COMMENTED** [review]
Looks like things get a little worse - but not too much.

I think actually `https://github.com/sofroniewn/napari/pull/15` was the better approach - because your PR is to my PR, whereas this PR is to my main. Instead for your other PR I think you just need to start from a different point in git history. One thing to try is to just branch off my PR and add the extra file - that should do it. Maybe @potating-potato can help you

File: napari/benchmarks/benchmark_qt_viewer_vectors.py (hunk 1 above)
Comment (kcpevey) [reply napari/benchmarks/benchmark_qt_viewer_vectors.py:27]: yep, can do

**Comment by kcpevey** [comment]
@sofroniewn Is this git approach what you're looking for? https://github.com/sofroniewn/napari/pull/17

**Comment by liu-ziyang** [comment]
~~curious why is this PR closed? is there a git problem or?~~

nvm I see this is superseded by #17

**Comment by kcpevey** [comment]
I was continuing @sofroniewn 's work [here](https://github.com/napari/napari/pull/4846) but this PR was merging back into main and also branched from his fork at the wrong hash. He requested fix it so I opened [this one](https://github.com/sofroniewn/napari/pull/17)
